packages/contracts-bedrock/src/universal/OptimismMintableERC20FactoryInterop.sol

```diff
@@ -91,12 +93,17 @@ contract OptimismMintableERC20FactoryInterop is OptimismMintableERC20Factory {
         emit HashOnionUpdated(_startingInnerLayer, _currentHashOnion);
     }
 
-    /// @notice One-time setter for the hashOnion value to be called by the ProxyAdmin.
+    /// @notice One-time setter for the hashOnion value to be called by the ProxyAdmin owner, triggered from L1.
     /// @param _hashOnion The new hashOnion value.
     function setHashOnion(bytes32 _hashOnion) external {
-        // TODO: Research if this is the correct address to check for.
-        if (msg.sender != Ownable(Predeploys.PROXY_ADMIN).owner()) revert Unauthorized();
         if (hashOnion() != 0) revert HashOnionAlreadySet();
+        if (
+            msg.sender != Predeploys.L2_CROSS_DOMAIN_MESSENGER
+                || ICrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER).xDomainMessageSender()
+                    != AddressAliasHelper.undoL1ToL2Alias(Ownable(Predeploys.PROXY_ADMIN).owner())
+        ) {
+            revert Unauthorized();
+        }
```

Comment:
Event though they are using the same `Unauthorized` error, i would divide them in 2 different checks for more readability

Reply:
Totally, i mean using only one custom error but with two different `if` blocks

Reply:
ah perfect. Makes sense yes